src/library/scala/collection/TraversableLike.scala

```diff
@@ -451,17 +451,23 @@ trait TraversableLike[+A, +Repr] extends Any
   }
 
   def groupBy[K](f: A => K): immutable.Map[K, Repr] = {
```

Comment:
I think that we can do better for small collections, so focussing on reducing the allocation cost for small collections and groups , and special casing LinearSeq I come up with something like
```
 def groupBy[K](f: A => K): immutable.Map[K, Repr] = {
    this.seq match {
      case //things with sensible isEmpty and "common" type
        _: LinearSeq[A] |
        _: IndexedSeq[A] |
        _: mutable.Set[A] |
        _: immutable.Set[A]
        if (isEmpty) =>
        //fast path, and avoid additional object creation
        Map.empty
      case _ =>
        val m = new mutable.HashMap[K, Builder[A, Repr]]()
        object grouper extends AbstractFunction1[A, Unit] with Function0[Builder[A, Repr]] {
          override def apply(): mutable.Builder[A, Repr] = newBuilder
          override def apply(elem: A): Unit =
            (m.getOrElseUpdate(f(elem), apply)) += elem
          @tailrec def applyAll(elems: LinearSeq[A]): Unit = {
            if (!elems.isEmpty) {
              apply(elems.head)
              applyAll(elems.tail)
            }
          }
        }
        this.seq match {
          case ls: LinearSeq[A] =>
            grouper.applyAll(ls)
          case _                =>
            this foreach grouper
        }
        m.size match {
          case 0 => Map.empty
          case 1 =>
            //coule be optimised
            val e1 = m.entriesIteratorPrivate.next()
            new immutable.Map.Map1(e1.key, e1.value.result())
          case 2 =>
            val it = m.entriesIteratorPrivate
            val e1 = it.next()
            val e2 = it.next()
            new immutable.Map.Map2( //
                                    e1.key, e1.value.result(), //
                                    e2.key, e2.value.result()
                                    )
          case 3 =>
            val it = m.entriesIteratorPrivate
            val e1 = it.next()
            val e2 = it.next()
            val e3 = it.next()
            new immutable.Map.Map3( //
                                    e1.key, e1.value.result(), //
                                    e2.key, e2.value.result(), //
                                    e3.key, e3.value.result()
                                    )
          case 4 =>
            val it = m.entriesIteratorPrivate
            val e1 = it.next()
            val e2 = it.next()
            val e3 = it.next()
            val e4 = it.next()
            new immutable.Map.Map4( //
                                    e1.key, e1.value.result(), //
                                    e2.key, e2.value.result(), //
                                    e3.key, e3.value.result(), //
                                    e4.key, e4.value.result()
                                    )
          case _ =>
            val it = m.entriesIteratorPrivate
            val m1 = immutable.HashMap.newBuilder[K, Repr]
            while (it.hasNext) {
              val entry = it.next()
              m1.+=((entry.key, entry.value.result()))
            }
            m1.result()
        }
  }
```
this still creates an additional Tuple2 to add to the target HashMap, and recalculates the hashcode of the key
The first part of this is reasonably easy to get around, but I cant see how the to get at the hash from the `HashEntry` in the `mutable.HashMap`. In fact it seems that  `mutable.HashMap` entries should contain a hashcode to reduce the calls to equals, like `java.util.HashMap.Node` does, but that seems outside the scope of this PR for sure

Reply:
pushed the code or this and the addOne rework for tuples to https://github.com/mkeskells/scala/tree/topic/group-by-2.12.x - just the last commit is the change